pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -87,6 +87,7 @@ static int _read(struct dtls_context_t *ctx, session_t *session, uint8_t *buf,
     sock->buffer.session = session;
 #ifdef SOCK_HAS_ASYNC
     if (sock->async_cb != NULL) {
+        memcpy(&sock->async_cb_session, session, sizeof(session_t));
```

Comment:
The `memcpy` seems unnecessary if the sessions are going to be used in the handler anyways because the handler is called directly after by `sock->async_cb`, so we don't have to worry about the memory of `session` going out of scope. A pointer would be enough. 

If the server wants to keep that information session after the handler exits, only then do the session need to be copied but that should be done by the application (dtls-client/server) and not `sock_dtls`.

Reply:
Though by not copying the session into the sock, `sock_dtls_get_event_session` would only be usable withing the async handler, as outside of it there is no guarantee that the pointer is valid and therefore should set to `NULL` when exiting the handler. This behavior should still solve the issue you're having in #15517 where you cannot access the session from the handler. Is there any uses for `sock_dtls_get_event_session` outside of the handler?

Reply:
The problem I had with a pointer was that I cannot set the pointer to NULL after the callback call.
```
sock->async_cb_session = session;
sock->async_cb(sock, SOCK_ASYNC_MSG_RECV, sock->async_cb_arg);
sock->async_cb_session = NULL;
```
The session is already NULL again when I call `sock_dtls_get_event_session` within the DTLS handler. 

No, `sock_dtls_get_event_session` should have no use case outside of the handler. It's not even guaranteed that you get a proper information out of it outside the handler.

Reply:
I have a branch with the pointer solution [here](https://github.com/janosbrodbeck/RIOT/tree/pr/dtls/get_event_session_pointer). I would prefer a solution with just a pointer. 

Or did I misunderstood you? :)

Reply:
the code in your branch is exactly how I imagined it. Weird that the session is already NULL though..

Reply:
Okay apparently I misunderstood how `sock_async` works. Although it seems that `sock->async_cb` is called inside of `_event`/`_read`, it is actually executed after leaving those functions. I discovered this after adding some printf. Can you confirm this @miri64?

This means, we accessed the session after we set it to `NULL`, causing it to segfault. So I tested without setting `async_cb_session` to `NULL` and it worked! Though I'm not sure if this have any other side effects..

Here is the log of the working test with more printf added:
```
main(): This is RIOT! (Version: 2021.01-devel-1872-gfdb01-pr/dtls/get_event_session_pointer)                                                                                                                       
DTLS sock example application                                                                                                                                                                                      
All up, running the shell now                                                                                                                                                                                      
> dtlss start                                                                                                                                                                                                      
dtlss start                                                                                                                                                                                                        
> entering _event                                                                                                                                                                                                  
leaving _event                                                                                                                                                                                                     
entering _dtls_handler                                                                                                                                                                                             
event: SOCK_ASYNC_CONN_RECV
Session handshake received: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
New client connected
leaving _dtls_handler
entering _dtls_handler
event: SOCK_ASYNC_CONN_RDY
Session became ready: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
leaving _dtls_handler
entering _read
leaving _read
entering _dtls_handler
event: SOCK_ASYNC_MSG_RECV
Message received: done mesg recevd
Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4
Received 12 bytes -- (echo)
leaving _dtls_handler
entering _dtls_handler
leaving _dtls_handler
entering _event
leaving _event
entering _dtls_handler
event: SOCK_ASYNC_CONN_FIN
Session was destroyed by peer: Peer: FE:80:00:00:00:00:00:00:E8:BC:C0:FF:FE:06:47:8B: Port: 12345 Family: 4

leaving _dtls_handler

```

Reply:
Ahh was wondering too when this gets called.
The problem I saw with not setting it to NULL is that it points to "random" memory and could potentially be misused?

Reply:
> Okay apparently I misunderstood how `sock_async` works. Although it seems that `sock->async_cb` is called inside of `_event`/`_read`, it is actually executed after leaving those functions. I discovered this after adding some printf. Can you confirm this @miri64?

Mhh there are actually two callbacks, that's maybe where the confusion is coming from. `sock->async_cb` is the asynchronous callback internal to the `sock_async` backend and then there is the frontend's callback (since `sock_async_event` also uses callbacks) which is called from the event handler (to make things portable between different frontend approaches including a bare metal approach of just using the backend they have the same signature). `sock->async_cb` is called from within `_event` and `_read` that is correct, but it merely triggers the frontend event (`event_post()` in the `sock_async_event` case). The actual event handling is done within the the event handler of the frontend, which will be (comparably) much later then the call to `sock->async_cb` and in a different thread (the medium priority thread of `event_thread` in `tests/pkg_tinydtls_sock_async`'s case. Hope that clears things

TL;DR passing pointers around breaks the asynchronous and decoupled nature of `sock_async`. @janosbrodbeck was right in copying the session instead.

Reply:
Question is: what do you do, if there were multiple events fired before the handler comes to handling it. With your approach you will always only have the chance to get the session of the last event fired (which given that events are identified by a bitfield you can't determined which it was).

Reply:
Thanks for the explanation. That makes the process more understandable. Will have to look at this mechanism in detail the next days...that's something you should always have in mind when using async callbacks.

> Question is: what do you do, if there were multiple events fired before the handler comes to handling it. With your approach you will always only have the chance to get the session of the last event fired (which given that events are identified by a bitfield you can't determined which it was).

I also thought of this while reading. The approach here then no longer works. The other idea with another parameter in the callback does not work then either. Hmm...

Reply:
This is about getting the session for very dedicated event types like `SOCK_ASYNC_CONN_FIN` or `SOCK_ASYNC_CONN_RDY`, right? Those are pretty rare and won't probably fire twice on the same `sock`. How about only setting the sessions for occurences of that event and specifically document that. For `SOCK_ASYNC_MSG_RECV` a user should use `sock_dtls_recv` anyway.

Reply:
Only those two events, yes. The number of sessions that a device can hold will probably never be huge. Sessions are expensive (even more so with wolfSSL). I also think it is very unlikely that this scenario will occur in practice. But i also don't want to exclude it to be honest. However, when it happens you can actually prevent resulting problems at the application level. It's not optimal...but I don't know any other solution at the moment. 

In case of doubt if problems occur we could make setting both event types configurable. Which means: If you only need one event type you could disable the setting of the other event type. This should lower the chance of the problem even more since then the session will be set for only one event type. For example I don't need the `SOCK_ASYNC_CONN_RDY` for gcoap currently.